miasm2/ir/ssa.py

```diff
@@ -0,0 +1,279 @@
+            else:
+                instructions.append(aff)
+
+        return instructions
+
+    @staticmethod
+    def _convert_block(ib, l):
+        """
+        Transforms an IRA block inplace into SSA
+        :param ib: IRA block to be transformed
+        :param l: list of SSA expressions
+        """
+        # iterator over SSA expressions
+        ssa_iter = iter(l)
+        # walk over IR blocks' assignblocks
+        for index, assignblk in enumerate(ib.irs):
+            # list of instructions
+            instructions = []
+            # insert SSA instructions
+            for dst in assignblk:
+                instructions.append(ssa_iter.next())
+            # replace instructions of assignblock in IRA block
+            ib.irs[index] = AssignBlock(instructions)
+
+    def _copy_block(self, block):
+        """
+        Returns a copy on an IRA block
+        :param block: asm_label
+        :return: IRA block
+        """
+        # retrieve IRA block
+        ib = self.ira.get_bloc(block)
+
+        # copy IRA block
+        irs = [assignblk.copy() for assignblk in ib.irs]
+        ib_ssa = irbloc(ib.label, irs)
+
+        # set next block
+        ib_ssa.dst = ib.dst.copy()
+
+        # add to SSA blocks dict
+        self.blocks.update({ib_ssa.label: ib_ssa})
+
+        return ib_ssa
+
+    def _rename_expressions(self, block):
+        """
+        Transforms variables and expressions
+        of an IRA block into SSA.
+
+        IR representations of an assembly instruction are evaluated
+        in parallel. Thus, RHS and LHS instructions will be performed
+        separately.
+        :param block: IRA block label
+        """
+        # list of IRA block's SSA expressions
+        ssa_expressions_block = []
+
+        # retrieve IRA block
+        ib = self.get_block(block)
+
+        # iterate block's IR expressions
+        for assignblk in ib.irs:
+            # list of parallel instructions
+            instructions = self._parallel_instructions(assignblk)
+            # list for transformed RHS expressions
+            rhs = []
+
+            # transform RHS
+            for e in instructions:
+                src = e.src
+                src_ssa = self._transform_expression_rhs(src)
+                # save transformed RHS
+                rhs.append(src_ssa)
+
+            # transform LHS
+            for e in instructions:
+                dst = e.dst
+                dst_ssa = self._transform_expression_lhs(dst)
+
+                # retrieve corresponding RHS expression
+                src_ssa = rhs.pop(0)
+
+                # rebuild SSA expression
+                e = m2_expr.ExprAff(dst_ssa, src_ssa)
+                self.expressions[dst_ssa] = src_ssa
+
+                # append ssa expression to list
+                ssa_expressions_block.append(e)
+
+        # replace blocks IR expressions with corresponding SSA transformations
+        self._convert_block(ib, ssa_expressions_block)
```

Comment:
Instead of having a flat list of instructions in `ssa_expressions_block` and iterating again on each destination of each `AssignBlock`, you can move the end of `_convert_block` into the above `for`.
It also seems easier to debug / understand for me, avoiding the fake separation between instructions.